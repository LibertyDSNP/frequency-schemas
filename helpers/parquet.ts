/* eslint-disable jest/no-export */
import fs from "fs";
import { ParquetSchema, ParquetWriter } from "@dsnp/parquetjs";
import { FrequencyParquetType, ParquetModel } from "../types/frequency";

type BloomFilterColumn = { column: string };
type RowGenerator = () => Record<string, unknown>;

const convertColumnType = (columnType: FrequencyParquetType): string => {
  if (typeof columnType === "string") {
    switch (columnType) {
      case "STRING":
        return "UTF8";
      default:
        return columnType;
    }
  }
  if ("INTEGER" in columnType) {
    // ParquetJs uses the old format still
    return `${columnType.INTEGER.sign ? "" : "U"}INT_${columnType.INTEGER.bit_width}`;
  }
  throw new Error("Not currently supported in transformation");
};

// Transform a ParquetModel (an array of ParquetColumn) to a ParquetSchema
const transformToParquetjs = (frequencySchema: ParquetModel): [ParquetSchema, BloomFilterColumn[]] => {
  const schema = Object.fromEntries(
    frequencySchema.map((column) => {
      return [
        column.name,
        {
          type: convertColumnType(column.column_type),
          compression: column.compression,
          statistics: false,
        },
      ];
    })
  );
  const bloomFilters = frequencySchema.reduce<BloomFilterColumn[]>((acc, x) => {
    if (x.bloom_filter) acc.push({ column: x.name });
    return acc;
  }, []);

  /* eslint-disable @typescript-eslint/no-explicit-any */
  return [new ParquetSchema(schema as any), bloomFilters];
};

export const testParquetSchema = async (model: ParquetModel) => {
  test("can build a parquet file", async () => {
    await expect(async () => {
      const [schema, bloom] = transformToParquetjs(model);
      await ParquetWriter.openStream(
        schema,
        {
          write: jest.fn(),
          end: jest.fn(),
        },
        { bloomFilters: bloom }
      );
    }).not.toThrow();
  });
};

export const testCompression = async (name: string, model: ParquetModel, rowGenerator: RowGenerator) => {
  test("Compressing all columns is best", async () => {
    const defaultSize = await generateParquetTestFileSize(name, model, 100, rowGenerator);

    for (const idx in model) {
      const testSchema = [...model];
      testSchema[idx].compression = "UNCOMPRESSED";
      const uncompressedSize = await generateParquetTestFileSize(name, testSchema, 100, rowGenerator);
      expect(defaultSize).toBeLessThan(uncompressedSize);
    }
  });
};

const generateParquetTestFileSize = async (
  name: string,
  rawSchema: ParquetModel,
  count: number,
  rowGenerator: RowGenerator
): Promise<number> => {
  const [schema, bloom] = transformToParquetjs(rawSchema);
  const path = `./test-${name}-size.parquet`;
  const writer = await ParquetWriter.openFile(schema, path, { bloomFilters: bloom });

  for (let i = 0; i < count; i++) {
    await writer.appendRow(rowGenerator());
  }
  await writer.close();
  const size = fs.statSync(path).size;
  fs.rmSync(path);
  return size;
};
