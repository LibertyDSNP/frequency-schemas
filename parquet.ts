import { ParquetSchema } from "@dsnp/parquetjs";
import type { ParquetType, FieldDefinition, SchemaDefinition, WriterOptions } from "@dsnp/parquetjs/dist/lib/declare";
import type { createSBBFParams } from "@dsnp/parquetjs/dist/lib/bloomFilterIO/bloomFilterWriter";
import { FrequencyParquetSchema, FrequencyParquetType, ParquetColumn } from "./types/frequency.js";
import * as dsnp from "./dsnp/index.js";

/**
 * All supported types from Parquetjs
 */
const supportedTypes = new Set([
  "BOOLEAN",
  "INT32",
  "INT64",
  "INT96",
  "FLOAT",
  "DOUBLE",
  "BYTE_ARRAY",
  "FIXED_LEN_BYTE_ARRAY",
  "UTF8",
  "MAP",
  "LIST",
  "ENUM",
  "DECIMAL",
  "DATE",
  "TIME_MILLIS",
  "TIME_MICROS",
  "TIMESTAMP_MILLIS",
  "TIMESTAMP_MICROS",
  "UINT_8",
  "UINT_16",
  "UINT_32",
  "UINT_64",
  "INT_8",
  "INT_16",
  "INT_32",
  "INT_64",
  "JSON",
  "BSON",
  "INTERVAL",
]);

/**
 * Simple check to make sure that type is supported.
 */
const isColumnTypeSupported = (incoming: string): incoming is ParquetType => {
  return supportedTypes.has(incoming);
};

/**
 * Error to capture all the unsupported edge cases
 */
export class UnsupportedFrequencySchemaError extends Error {
  constructor(msg: string) {
    const message = `Unsupported Frequency schema: ${msg}`;
    super(message);
    this.name = "UnsupportedFrequencySchemaError";
  }
}

const convertColumnType = (columnType: FrequencyParquetType): FieldDefinition["type"] => {
  if (typeof columnType === "string") {
    if (columnType === "STRING") return "UTF8";
    if (isColumnTypeSupported(columnType)) return columnType;
    throw new UnsupportedFrequencySchemaError(columnType.toString());
  }
  // ParquetJs uses the old format still, so not all options are available
  if ("INTEGER" in columnType) {
    return `${columnType.INTEGER.sign ? "" : "U"}INT_${columnType.INTEGER.bit_width}` as ParquetType;
  }
  if ("TIMESTAMP" in columnType && columnType.TIMESTAMP.is_adjusted_to_utc && columnType.TIMESTAMP.unit !== "NANOS") {
    return `TIMESTAMP_${columnType.TIMESTAMP.unit}` as ParquetType;
  }
  if ("TIME" in columnType && columnType.TIME.is_adjusted_to_utc && columnType.TIME.unit !== "NANOS") {
    return `TIME_${columnType.TIME.unit}` as ParquetType;
  }

  throw new UnsupportedFrequencySchemaError(columnType.toString());
};

/**
 * Converts a field from a JSON Schema into a Parquet Field Definition
 */
const fromColumn = (column: ParquetColumn): FieldDefinition => {
  return {
    type: convertColumnType(column.column_type),
    compression: column.compression,
    statistics: false,
  };
};

/**
 * Converts supported Json Schemas into Parquet Schema Definitions
 */
const toSchema = (frequencySchema: FrequencyParquetSchema): SchemaDefinition => {
  const schema: SchemaDefinition = {};

  for (const column of frequencySchema) {
    schema[column.name] = fromColumn(column);
  }

  return schema;
};

/**
 * Create a new schema from a Frequency Parquet Schema (frequency.xyz)
 * Also provides the Writer Options as Frequency Schemas support bloom filter selection.
 */
export const fromFrequencySchema = (
  frequencySchema: FrequencyParquetSchema | dsnp.ParquetSchemaName
): [ParquetSchema, WriterOptions] => {
  if (typeof frequencySchema === "string") {
    const deploy = dsnp.getSchema(frequencySchema);
    if (!deploy || deploy.modelType !== "Parquet") throw new Error("Unknown schema");
    frequencySchema = deploy.model;
  }
  const schema: SchemaDefinition = toSchema(frequencySchema);
  const bloomFilters = frequencySchema.reduce<createSBBFParams[]>((acc, x) => {
    if (x.bloom_filter) acc.push({ column: x.name });
    return acc;
  }, []);
  return [
    new ParquetSchema(schema),
    {
      bloomFilters,
    },
  ];
};
