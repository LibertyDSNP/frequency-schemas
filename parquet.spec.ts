import fs from "fs";
import { ParquetWriter, ParquetReader } from "@dsnp/parquetjs";
import { fromFrequencySchema } from "./parquet.js";
import { broadcast } from "./dsnp/index.js";

describe("Frequency Schema Conversion Test File", () => {
  const [parquetSchema, writerOptions] = fromFrequencySchema(broadcast);

  const row1 = {
    announcementType: 2,
    contentHash: "0x12345678",
    fromId: 12n,
    url: "https://github.com/LibertyDSNP/parquetjs/",
  };

  const path = "./frequency-schema-test-file.parquet";

  let reader: ParquetReader;

  beforeAll(async () => {
    const writer = await ParquetWriter.openFile(parquetSchema, path, writerOptions);
    writer.appendRow(row1);
    await writer.close();

    reader = await ParquetReader.openFile(path);
  });

  afterAll(async () => {
    await reader.close();
    fs.rmSync(path);
  });

  it("schema is generated correctly", () => {
    expect(parquetSchema).toMatchSnapshot();
  });

  it("schema is encoded correctly", () => {
    expect(reader.metadata?.schema).toMatchSnapshot();
  });

  it("output matches input", async () => {
    const cursor = reader.getCursor();
    const row = await cursor.next();
    const rowData = {
      ...row1,
      contentHash: Buffer.from([48, 120, 49, 50, 51, 52, 53, 54, 55, 56]),
    };
    expect(row).toEqual(rowData);
  });
});
