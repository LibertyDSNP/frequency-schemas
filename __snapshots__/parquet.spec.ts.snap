// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frequency Schema Conversion Test File schema is encoded correctly 1`] = `
[
  {
    "converted_type": null,
    "field_id": null,
    "logicalType": null,
    "name": "root",
    "num_children": 4,
    "precision": null,
    "repetition_type": null,
    "scale": null,
    "type": null,
    "type_length": null,
  },
  {
    "converted_type": 17,
    "field_id": null,
    "logicalType": null,
    "name": "announcementType",
    "num_children": null,
    "precision": null,
    "repetition_type": 0,
    "scale": null,
    "type": 1,
    "type_length": null,
  },
  {
    "converted_type": null,
    "field_id": null,
    "logicalType": null,
    "name": "contentHash",
    "num_children": null,
    "precision": null,
    "repetition_type": 0,
    "scale": null,
    "type": 6,
    "type_length": null,
  },
  {
    "converted_type": 14,
    "field_id": null,
    "logicalType": null,
    "name": "fromId",
    "num_children": null,
    "precision": null,
    "repetition_type": 0,
    "scale": null,
    "type": 2,
    "type_length": null,
  },
  {
    "converted_type": 0,
    "field_id": null,
    "logicalType": null,
    "name": "url",
    "num_children": null,
    "precision": null,
    "repetition_type": 0,
    "scale": null,
    "type": 6,
    "type_length": null,
  },
]
`;

exports[`Frequency Schema Conversion Test File schema is generated correctly 1`] = `
ParquetSchema {
  "fieldList": [
    {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "announcementType",
      "originalType": "INT_32",
      "path": [
        "announcementType",
      ],
      "precision": undefined,
      "primitiveType": "INT32",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
    {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "contentHash",
      "originalType": undefined,
      "path": [
        "contentHash",
      ],
      "precision": undefined,
      "primitiveType": "BYTE_ARRAY",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
    {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "fromId",
      "originalType": "UINT_64",
      "path": [
        "fromId",
      ],
      "precision": undefined,
      "primitiveType": "INT64",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
    {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "url",
      "originalType": "UTF8",
      "path": [
        "url",
      ],
      "precision": undefined,
      "primitiveType": "BYTE_ARRAY",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
  ],
  "fields": {
    "announcementType": {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "announcementType",
      "originalType": "INT_32",
      "path": [
        "announcementType",
      ],
      "precision": undefined,
      "primitiveType": "INT32",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
    "contentHash": {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "contentHash",
      "originalType": undefined,
      "path": [
        "contentHash",
      ],
      "precision": undefined,
      "primitiveType": "BYTE_ARRAY",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
    "fromId": {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "fromId",
      "originalType": "UINT_64",
      "path": [
        "fromId",
      ],
      "precision": undefined,
      "primitiveType": "INT64",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
    "url": {
      "compression": "GZIP",
      "dLevelMax": 0,
      "encoding": "PLAIN",
      "name": "url",
      "originalType": "UTF8",
      "path": [
        "url",
      ],
      "precision": undefined,
      "primitiveType": "BYTE_ARRAY",
      "rLevelMax": 0,
      "repetitionType": "REQUIRED",
      "scale": undefined,
      "statistics": false,
      "typeLength": undefined,
    },
  },
  "schema": {
    "announcementType": {
      "compression": "GZIP",
      "encoding": "PLAIN",
      "statistics": false,
      "type": "INT_32",
    },
    "contentHash": {
      "compression": "GZIP",
      "encoding": "PLAIN",
      "statistics": false,
      "type": "BYTE_ARRAY",
    },
    "fromId": {
      "compression": "GZIP",
      "encoding": "PLAIN",
      "statistics": false,
      "type": "UINT_64",
    },
    "url": {
      "compression": "GZIP",
      "encoding": "PLAIN",
      "statistics": false,
      "type": "UTF8",
    },
  },
}
`;
